### DISCUSSION TITLE
Test nicht erfolgreich

### DISCUSSION THREAD

**Discussion by njoppen** [Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II#5] [Show and tell]
Ich habe eine SD-Karte generiert, es kommt ein Fehler.

![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/1ecb849e-29d8-43b4-b953-5090a3d2637d)

Der Code von 0600.07FF ist im Forum, da ich hier keine Datei anhängen kann. Hier eine Kopie:

0600- B0 3C 90 4A 43 4F 53 20  20 20 20 00 02 08 04 10    .<.JCOS    .....
0610- 02 00 00 00 00 F8 00 00  3F 00 FF 00 00 08 00 00    ........?.......
0620- C3 14 20 00 FE 07 00 00  00 00 00 00 02 00 00 00    .. .............
0630- 01 00 06 00 00 00 00 00  00 00 00 00 00 00 A5 FD    ................
0640- 8D 08 04 AD 0D 06 8D 0B  04 AD 15 06 8D 09 04 C9    ................
0650- F8 F0 10 AD 13 06 8D 04  04 AD 14 06 8D 05 04 A2    ................
0660- 02 D0 02 A2 01 8E 0A 04  A2 00 A0 04 20 59 07 AE    ............ Y..
0670- 0E 06 AC 0F 06 20 86 07  A2 14 A0 04 20 77 07 AE    ..... ...... w..
0680- 16 06 AC 17 06 20 86 07  A2 18 A0 04 20 77 07 AC    ..... ...... w..
0690- 10 06 8C 0C 04 88 D0 03  4C 48 07 20 90 07 88 D0    ........LH. ....
06A0- FA A2 1C A0 04 20 77 07  AE 11 06 8E 0D 04 AC 12    ..... w.........
06B0- 06 8C 0E 04 86 FC 84 FD  A2 04 46 FD 66 FC CA D0    ..........F.f...
06C0- F9 86 FE 86 FF 20 90 07  A2 10 A0 04 20 77 07 A9    ..... ...... w..
06D0- 00 85 DC 85 EA A9 02 85  DD 85 EB A2 1C A0 04 20    ...............
06E0- 62 F4 A0 00 B1 EA F0 60  B1 EA C9 20 90 49 C9 61    b......`... .I.a
06F0- 90 02 29 DF D9 D3 07 D0  3E C8 C0 0B 90 EA A9 00    ..).....>.......
0700- 85 DC A9 20 85 DD 38 A0  1A B1 EA E9 02 85 E8 C8    ... ..8.........
0710- B1 EA E9 00 85 E9 AD 0B  04 4A F0 07 06 E8 26 E9    .........J....&.
0720- 4C 19 07 A6 E8 A4 E9 20  86 07 A2 F8 A0 00 20 62    L...... ...... b
0730- F4 20 06 E0 4C 00 20 18  A9 20 65 EA 85 EA A9 00    . ..L. .. e.....
0740- 65 EB 85 EB C9 04 90 9A  A9 A3 85 EA A9 07 85 EB    e...............
0750- 20 83 E0 20 47 E0 4C 00  E0 86 E8 84 E9 A0 03 B1     .. G.L.........
0760- E8 99 F8 00 88 10 F8 60  86 E8 84 E9 A0 03 B1 E8    .......`........
0770- 99 FC 00 88 10 F8 60 86  E8 84 E9 A0 03 B9 F8 00    ......`.........
0780- 91 E8 88 10 F8 60 86 FC  84 FD A2 00 86 FE 86 FF    .....`..........
0790- 18 A2 00 08 28 B5 FC 75  F8 95 F8 08 E8 E0 04 D0    ....(..u........
07A0- F3 28 60 0D 0D 20 4E 6F  20 76 61 6C 69 64 20 62    .(`.. No valid b
07B0- 6F 6F 74 20 64 69 73 6B  0D 20 50 72 65 73 73 20    oot disk. Press
07C0- 61 6E 79 20 6B 65 79 20  74 6F 20 72 65 73 74 61    any key to resta
07D0- 72 74 00 43 50 4D 4A 43  20 20 20 49 4D 47 FF FF    rt.CPMJC   IMG..
07E0- FF FF FF FF FF FF FF FF  FF FF FF FF FF FF FF FF    ................

**Comment by Dietrich-L** [comment]
Wir suchen weiter

**Comment by Dietrich-L** [comment]
Funktioniert denn boot.sys bei dir?

**Comment by njoppen** [comment]
Meinst du die Version von Jörg? Ja, die läuft mit seiner BootCode_FAT16.bin.

**Comment by Dietrich-L** [comment]
Ok, dann versuch bitte folgendes:
Benenne meine CPMJC.IMG in BOOT.SYS um und kopiere diese auf eine neu präparierte SD mit Jörgs  .bin. Grund: CPM-65 braucht seine Images in einer Kette zusammenhängender Sektoren auf der SD.
Wenn das auch nicht klappt, dann kopiere mal die umbenannte BOOT.SYS auf Jörgs BOOT.SYS auf der funktionierenden SD. Damit sollte es mindestens zu einer Bootmeldung kommen, hoffentlich sogar zum Prompt A>

Viel Glück
Dietrich

**Comment by njoppen** [comment]
Der erste Vorschlag hat zu nichts geführt. Hier der zweite:
![BOOT failed](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/073d6e4a-d7a0-453b-a65d-42c677eba33a)
Diese Fehlermeldung kommt, dann Reset und von vorne, endlos.

**Comment by njoppen** [comment]
Nur mal so: du weißt, dass der Junior ][ nur den Original-Monitor des ursprünglichen Juniors hat und nicht die original TM- oder PM -ROMs? Solltest du daraus Routinen nutzen, kann das natürlich nicht funktionieren.

**Comment by Dietrich-L** [comment]
Ja, das weis ich. Mein Junior hat mittlerweile keine Original-Software mehr. Nur die Monitorroutinen für das 7-Segment Display sind noch wiederzuerkennen. Das ist es also nicht.
Deine Rückmeldung ist erstmal hilfreich - du bist erstmals in meinem Code gelandet. Das bedeuted, dass mindestens der Bootsector meines Images geladen wurde und zwar nach $2000 und daß die Ausgabe von Zeichen über die entsprechende ROM-Routine des JC-II geklappt hat. Meine Error-Routine springt dann die Reset-Routine im JC-II-ROM an. Das funktioniert also auch. Nun ist zu klären, warum mein Bootprozess nicht läuft.
Ich mache mir Gedanken.

**Comment by Dietrich-L** [comment]
Das Ergebnis zeigt allerdings auch, dass mit der anderen SD etwas nicht stimmt. Die müsste sich nämlich identisch verhalten.

**Comment by Dietrich-L** [comment]
Ok, ein schneller Blick in meine BOOT.ASM offenbart gleich mehrere Fehler. Ich bin morgen und übermorgen verreist - es wird also ein paar Tage dauern. Keine große Sache, aber ich brauche den Kopf frei…

**Comment by Dietrich-L** [comment]
Ich habe eben eine BOOT.SYS und eine CPMJC.IMG mit korrigiertem Bootsector abgelegt. Sollte nun etwas besser gehen.

**Comment by njoppen** [comment]
Mit den neuen Files ist das Ergebnis
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/5cedc228-e221-48b8-b3c0-e4108d21b2af)

**Reply by Dietrich-L** [reply]
Das ist unerwartet und ein Rückschritt. Mein Code wird weder geladen noch ausgeführt. Da waren wir schon weiter.
Wir brauchen als Basis eine SD, die wenigstens Jörgs BOOT.SYS bootet. Auf dieser SD ersetzt du dann Jörgs BOOT.SYS durch meine - einfach drüberkopieren. Dann muss sich etwas ändern.

**Comment by njoppen** [comment]
Bin so vorgegangen.
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/d9b6bfb4-8583-41f5-8e47-abd11b8b901b)

**Reply by Dietrich-L** [reply]
Super - du bist beim CPM-65 Prompt angekommen. In CPM-65 gibt es keine residenten Befehle, nur Programm, die man aufruft. Das Directoryprogramm heißt D.COM. Also am Prompt D eingeben und schon siehst du, was auf dem Image ist - hoffentlich 🙏

**Comment by njoppen** [comment]
Leider nicht nicht ganz....
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/d9175807-b5dd-4170-aa73-4ad133a6fdae)

**Comment by njoppen** [comment]
Auf der Karte ist das CPMJC.IMG vorhanden.

**Comment by Dietrich-L** [comment]
Ich prüfe das. D.COM wird offensichtlich geladen und gestartet, bringt die Startmeldung und stürzt dann ab - möglicherweise beim Zugriff auf das Directory. Da es noch einen bekannten Fehler im BIOS gibt, und zwar bei der Umrechnung von relativen Sectornummern aus CPM-65 in absolute Sectornummern auf der SD, möchte ich erst den beheben und dann weitere Versuche machen. Was du auf die Schnelle noch tun könntest, wäre D A:. Du solltest dann eine Aufforderung zum Einlegen von Diskette A: bekommen, die du mit Tastendruck quittierst. Dann erfolgt vermutlich der Absturz. Bitte mach wieder einen Screenshot.

Wenn du ungeduldig bist, kannst du aber auch weitere Programme auf dem Image testen. Zu nennen wären da DUTIL (Disketten-Sector-Editor), Alloc, Sysinfo, Type d.asm und FORTH-E und berichten, ob etwas davon geht.

Viel Erfolg
Dietrich

**Comment by Dietrich-L** [comment]
Ok - nächster Versuch. BOOT.SYS und CPMJC.IMG sind korrigiert. Nun müsste D.COM laufen.
Falls nicht, mache bitte zusätzlich zum Screenshot noch einen Speicherauszug 2000:2A00. Dann kann ich hoffentlich etwas sehen.
Ich habe auf dem Image nun auch eine erste Version von DEBUG.COM abgelegt. Aufruf DEBUG D.COM. dann hat du D.COM ab $2000 im RAM. Für die Kommandoübersicht von DEBUG am Prompt ein ? eingeben.

Viel Erfolg
Dietrich

**Comment by njoppen** [comment]
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/2f178776-e178-448b-83f4-4701995c0f70)
Voilà
[2000.2a00.txt](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/files/12243677/2000.2a00.txt)

**Comment by Dietrich-L** [comment]
Wir sind fast da...
Ich musste noch einen Fehler im BIOS korrigieren - das SD-TYPE Byte war ja nach $1A gerutscht... Das neue BIOS V0.4 ist online und die Images sind geupdated.
D.COM sollte nun funktionieren
DEBUG startet auch - der korrekte Aufruf ist aber 'DEBUG D.COM' Der Punkt ist wichtig ;-) Bitte beachten, dass der Aussprung aus DEBUG noch einen BUG hat, der dazu führt, dass das System neu gebootet werden muss. Ich arbeite noch dran.

Ich sehe mir noch den Speicherauszug an. Bitte probiert mal fleissig rum und dokumentiert, was so geht und nicht geht.
Wenn ihr Doku braucht, sagt Bescheid - ich gebe mir dann Mühe.

Gruß
Dietrich

**Comment by njoppen** [comment]
Klasse, das DEBUG funktioniert, habe einen Memory Dump damit erstellt!

**Reply by Dietrich-L** [reply]
Wenn DEBUG funktioniert, wäre das toll. Damit kannst du Breakpoints setzen und Programme im Single Step abarbeiten. Und natürlich alle Prozessor-Register ansehen. Debugger ist auch dabei. Und vieles mehr. Das ist mein Schweizer Taschenmesser zur Fehlersuche.

**Comment by njoppen** [comment]
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/5728181a-4a1d-4910-9d14-46cfb378ef9d)

**Reply by Dietrich-L** [reply]
Super - FORTH geht. SYSINFO muss ich noch auf den JC II anpassen
-Titel
-ROM Version

**Comment by njoppen** [comment]
DUTIL hängt bei mir, "DUTIL 1.5      DRIVE A:" und bleibt dort stecken

ALLOC   :
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/2c377410-359a-49da-a357-e8044a881dd3)

TYPE d.asm scheint zu funktionieren, ebenso FORTH-E.

Gruß Norbert

**Comment by Dietrich-L** [comment]
Bitte versuche nochmal D um das Directory auszugeben.
Bitte checke mit SYSINFO dass BIOS V0.4 aktiv ist.

**Comment by njoppen** [comment]
![image](https://github.com/Dietrich-L/CPM-65_for_JUNIOR_COMPUTER_II/assets/19375415/7f3c268a-faf3-4c7a-a138-50415b030ab6)

Ich habe noch Version 3.

**Reply by njoppen** [reply]
Bin heute nicht da.

**Comment by njoppen** [comment]
Bei mir ist  BIOS V0.3 aktiv, habe definitiv die Files aus der aktuellen (03.08., 16:05) ZIP-Datei entpackt und benutzt.
D hängt damit. Woher bekomme ich die BIOS V0.4 - Version?

**Reply by Dietrich-L** [reply]
Mein Fehler. Ich habe vergessen, die Version einzuspielen. Im Image von heute ist das korrigiert zusammen mit einem Fehler beim Sichern der SD_TYPE Variable.
Mit dieser Korrektur sollte DEBUG ohne Absturz verlassen werden.
Das Thema D.COM habe ich in einen Issue gepackt, da mir das immer noch ein Rätsel ist. Wir setzen das Testprogram bitte dort fort.

**Comment by Dietrich-L** [comment]
Bei näherer Überlegung bin ich darauf gekommen, dass einige meiner Programme wichtige Zeropage Adressen von Jörgs ROM überschreiben und ich nicht alle gesichert habe. Insbesondere fürchte ich, dass D.COM und auch DUTIL.COM $14/$15 überschreiben und damit die I/O abschiessen. Ich werde in den nächsten Tage eine BIOS-Version machen, die diese Speicherstellen sichert. Ist leider nicht trivial. Hatte ich beim Apple II Port auch...
